### PR TITLE
Fix glob bug.

### PR DESCRIPTION
**Pull request by Thom1729**
For #134.

After much research and testing, I determined that the fundamental bug was that every pattern component except for `**` should match at least one character — so `/A*B/` should match `AB`, but `A/*` shouldn't match `A/`. This behavior should match other glob engines and resolve the linked bug.

This behavior varies slightly from `sublime.find_resources('*')`, which will find all resources (like `sublime.find_resources('')`.

It's probably bad news for other reasons if `sublime.find_resources()` returns directories, but at least now it shouldn't be bad news for us in particular.

### CODE REVIEW

**Comment by Thom1729** [comment]
N.B. In order to directly test #134, we'd have to create a zipfile with directory entries, which I'm not sure how to do within Python.

**Comment by FichteFoll** [comment]
> N.B. In order to directly test #134, we'd have to create a zipfile with directory entries, which I'm not sure how to do within Python.

Yeah, I tried writing such a file with the zipfilemodule, but you can't create "files" with trailing slashes.
I suppose we need to bundle such a zip as an artifact.

File: st3/sublime_lib/_util/glob.py [hunk 1]
```diff
@@ -7,36 +7,45 @@
...
-        elif part == '?':
-            s += r'(?:[^/])'
-        elif part[0] == '[':
-            s += part
-        elif '**' in part:
+        elif component == '*':
+            # Component must not be empty.
+            expr_string += r'(?:[^/])+' + '/'
+        elif component == '**':
+            expr_string += r'(?:(?:.*)(?:\Z|/))?'
```
Comment (FichteFoll) [inline st3/sublime_lib/_util/glob.py:31]:
```suggestion
            expr_string += r'(?:.*(?:\Z|/))?'
```

File: tests/test_glob.py [hunk 2]
```diff
@@ -115,6 +115,7 @@ def test_recursive(self):
         self._test_matches(
             'Foo/**',
             [
+                'Packages/Foo/',
```
Comment (FichteFoll) [inline tests/test_glob.py:118]: Should we match this at all?

Comment (Thom1729) [reply tests/test_glob.py:118]: Yes, because a UNIX glob would and I don't think we need to deviate. It's a corner case that probably won't come up very often, but if it does then I think handling it like a UNIX glob is the least surprising thing to do.

**Comment by Thom1729** [comment]
> I suppose we need to bundle such a zip as an artifact.

Ugh. You're probably right, but still ugh. I've [opened a question](https://stackoverflow.com/questions/66970927/create-zipfile-with-directory-entries-in-python) on Stack Overflow looking for an alternative.

**Comment by Thom1729** [comment]
Do you think this can be merged as-is, punting the direct zipfile test to a future PR?

**Review by FichteFoll — COMMENTED** [review]
Yeah, we can merge this.
